laser_scan_densifier/src/laser_scan_densifier.cpp

```diff
@@ -0,0 +1,110 @@
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the CCNY Robotics Lab nor the names of its
+ *       contributors may be used to endorse or promote products derived from
+ *       this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "laser_scan_densifier/laser_scan_densifier.h"
+
+namespace scan_tools {
+
+LaserScanDensifier::LaserScanDensifier(ros::NodeHandle nh, ros::NodeHandle nh_private):
+  nh_(nh), 
+  nh_private_(nh_private)
+{
+  ROS_INFO ("Starting LaserScanDensifier");
+
+  // **** get paramters
+
+  if (!nh_private_.getParam ("step", step_))
+    step_ = 2;
+
+  ROS_ASSERT_MSG(step_ > 0, "step parameter is set to %, must be > 0", step_);
+
+  // **** advertise topics
+
+  scan_publisher_ = nh_.advertise<sensor_msgs::LaserScan>("scan_dense", 1);
+
+  // **** subscribe to laser scan messages
+
+  scan_subscriber_ = nh_.subscribe("scan", 1, &LaserScanDensifier::scanCallback, this);
+}
+
+LaserScanDensifier::~LaserScanDensifier ()
+{
+  ROS_INFO ("Destroying LaserScanDensifier");
+}
+
+void LaserScanDensifier::scanCallback (const sensor_msgs::LaserScanConstPtr& scan_msg)
+{
+  sensor_msgs::LaserScan::Ptr scan_dense;
+  scan_dense = boost::make_shared<sensor_msgs::LaserScan>();
+
+  // copy over equal fields
+
+  scan_dense->header          = scan_msg->header;
+  scan_dense->range_min       = scan_msg->range_min;
+  scan_dense->range_max       = scan_msg->range_max;
+  scan_dense->angle_min       = scan_msg->angle_min;
+  scan_dense->angle_max       = scan_msg->angle_max;
+  scan_dense->angle_increment = scan_msg->angle_increment / step_;
+  scan_dense->time_increment  = scan_msg->time_increment;
+  scan_dense->scan_time       = scan_msg->scan_time;
+
+  scan_dense->ranges.clear();
+  scan_dense->intensities.clear();
+
+  for (unsigned int i = 0; i < scan_msg->ranges.size()-1; i++)
+  {
+    scan_dense->ranges.insert(scan_dense->ranges.end(), step_, scan_msg->ranges[i]);
```

Comment:
Would it make more sense to interpolate all intermediate values by the average, median or Gaussian weighted average from the neighboring measurements from left and right instead of just copying a value to the right?

Reply:
I now support both modes...in our usecase...we apply this node onto processed data...where interpolation introduces false data points we don't want...